### PR TITLE
Add elasticache module for elasticache resource creation

### PR DESCRIPTION
**Pull request by mjcmtb**
Initial PR for a new Elasticache module that supports Memcached, Redis, and Redis Multi-Shard.

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+  tags = {
+    Environment     = "${var.environment}"
+    ServiceProvider = "Rackspace"
+  }
+}
+
+###<Common Resources>###
+
+resource "aws_elasticache_subnet_group" "elasticache_subnet_group" {
+  count      = "${local.conflict_exists ? 0 : 1}"
```
Comment (bohn002) [inline main.tf]: Is this conflict not caught by Plan?

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+  encryption_supported = "${lookup(local.elasticache_engine["${var.elasticache_engine_type}"], "encryption_supported")}"
+
+  # Determine if this qualifies as a redis multi shard instance
+  redis_multishard = "${local.elasticache_name == "redis" && var.redis_multi_shard ? true : false}"
+
+  # Default for number_of_nodes is 1, but at least 2 is required for redis non-multi-shard
+  redis_node_count = "${var.number_of_nodes < 2 ? 2 : var.number_of_nodes}"
+
+  # Construct cluster naming with cluster version here
+  substring_length         = "${19 - length(var.cluster_name_version) > length(var.cluster_name) ? length(var.cluster_name) : 19 - length(var.cluster_name_version)}"
```
[the remark this answers is not in the document]
Comment (mjcmtb) [reply main.tf]: I've added more comments around this and mentioned limits in the var file.

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+  # If redis multi shard and memcached is specified, consider this a conflict to prevent resources from being created.
+  conflict_exists = "${var.redis_multi_shard && local.elasticache_name == "memcached" ? true : false}"
+
+  # Set default port
+  default_port = "${local.elasticache_name == "memcached" ? "11211" : "6379"}"
+  set_port     = "${var.cache_cluster_port != "" ? var.cache_cluster_port : local.default_port}"
+
+  # Determine if Instance Class is T2 instance
+  instance_prefix = "${element(split(".",var.instance_class), 1)}"
+  is_t2           = "${local.instance_prefix == "t2" ? true : false}"
```
[the remark this answers is not in the document]
Comment (mjcmtb) [reply main.tf]: I've gone ahead comments on why this matters. Only t2 matters.

File: main.tf (hunk 1 above)
Comment (mjcmtb) [reply main.tf]: It does not get caught by plan. Usually conflicts are caught by plan if you provided conflicting parameters to a single resource. What this module does is it actually provisions several different types of "stacks" (coming from three separate CFN templates). The stack being provisioned is dependent on the inputs provided to the vars. If the user specifies a memcached engine and sets redis multi shard to true, then there is something wrong. If a conflict is seen, no resources should be built.

Comment (bohn002) [reply main.tf]: awesome , thanks.

**Pull request merged**